plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java

```diff
@@ -0,0 +1,570 @@
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.concurrent.MoreFutures.toListenableFuture;
+import static io.airlift.slice.SizeOf.estimatedSizeOf;
+import static io.trino.plugin.exchange.filesystem.FileSystemExchangeManager.PATH_SEPARATOR;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.lang.System.arraycopy;
+import static java.util.Objects.requireNonNull;
+import static java.util.Objects.requireNonNullElseGet;
+
+public class AzureBlobFileSystemExchangeStorage
+        implements FileSystemExchangeStorage
+{
+    private final int blockSize;
+    private final BlobServiceClient blobServiceClient;
+    private final BlobServiceAsyncClient blobServiceAsyncClient;
+
+    @Inject
+    public AzureBlobFileSystemExchangeStorage(ExchangeAzureConfig config)
+    {
+        requireNonNull(config, "config is null");
+        this.blockSize = toIntExact(config.getAzureStorageBlockSize().toBytes());
+        Optional<String> connectionString = config.getAzureStorageConnectionString();
+
+        BlobServiceClientBuilder blobServiceClientBuilder = new BlobServiceClientBuilder();
+        if (connectionString.isPresent()) {
+            blobServiceClientBuilder.connectionString(connectionString.get());
+        }
+        else {
+            blobServiceClientBuilder.credential(new DefaultAzureCredentialBuilder().build());
+        }
+
+        this.blobServiceClient = blobServiceClientBuilder.buildClient();
+        this.blobServiceAsyncClient = blobServiceClientBuilder.buildAsyncClient();
+    }
+
+    @Override
+    public void createDirectories(URI dir)
+            throws IOException
+    {
+        // Nothing to do for Azure
+    }
+
+    @Override
+    public ExchangeStorageReader createExchangeStorageReader(Queue<ExchangeSourceFile> sourceFiles, int maxPageStorageSize)
+    {
+        return new AzureExchangeStorageReader(blobServiceAsyncClient, sourceFiles, blockSize, maxPageStorageSize);
+    }
+
+    @Override
+    public ExchangeStorageWriter createExchangeStorageWriter(URI file, Optional<SecretKey> secretKey)
+    {
+        String containerName = getContainerName(file);
+        String blobName = getPath(file);
+        BlockBlobAsyncClient blockBlobAsyncClient = blobServiceAsyncClient
+                .getBlobContainerAsyncClient(containerName)
+                .getBlobAsyncClient(blobName)
+                .getBlockBlobAsyncClient();
+        if (secretKey.isPresent()) {
+            blockBlobAsyncClient = blockBlobAsyncClient.getCustomerProvidedKeyAsyncClient(new CustomerProvidedKey(secretKey.get().getEncoded()));
+        }
+        return new AzureExchangeStorageWriter(blockBlobAsyncClient, blockSize);
+    }
+
+    @Override
+    public boolean exists(URI file)
+            throws IOException
+    {
+        return blobServiceClient.getBlobContainerClient(getContainerName(file)).getBlobClient(getPath(file)).exists();
```

Comment:
Is `blobServiceClient.getBlobContainerClient` cheap? Do we need to cache it?

Reply:
It doesn't have any network ops. Just class construction, negligible cost?

Reply:
good advice

Reply:
If it's just a matter of creating a wrapper object than it's fine